// zip-build/lib/userResolveConflictAsync.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.userResolveConflictAsync = exports.appendTimestampToFilename = void 0;
var fs = require("fs");
var path = require("path");
var inquirer = require("inquirer");
var getTimestampString_1 = __importDefault(require("./getTimestampString"));
var Choices;
(function (Choices) {
    Choices["TIMESTAMP"] = "Rename output file appending the current timestamp";
    Choices["RENAME"] = "Rename output file with another name";
    Choices["OVERWRITE"] = "Overwrite existing file";
    Choices["EXIT"] = "Exit";
})(Choices || (Choices = {}));
function appendTimestampToFilename(filename) {
    var timestamp = (0, getTimestampString_1.default)();
    var _a = path.parse(filename), name = _a.name, ext = _a.ext;
    return "".concat(name, "_").concat(timestamp).concat(ext);
}
exports.appendTimestampToFilename = appendTimestampToFilename;
function userResolveConflictAsync(dstdir, filename) {
    return __awaiter(this, void 0, void 0, function () {
        var choices, MSG, QNAME, outfileName, _a, ANS_REN, OUTPATH, URI;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    choices = Object.values(Choices);
                    MSG = "The file '".concat(filename, "' already exists in directory ")
                        + "'".concat(dstdir, "'.. What do you want to do?");
                    return [4 /*yield*/, inquirer.prompt([{
                                type: 'list',
                                name: 'QNAME',
                                message: MSG,
                                choices: choices,
                            }])];
                case 1:
                    QNAME = (_b.sent()).QNAME;
                    _a = QNAME;
                    switch (_a) {
                        case Choices.TIMESTAMP: return [3 /*break*/, 2];
                        case Choices.RENAME: return [3 /*break*/, 3];
                        case Choices.OVERWRITE: return [3 /*break*/, 5];
                        case Choices.EXIT: return [3 /*break*/, 6];
                    }
                    return [3 /*break*/, 6];
                case 2:
                    outfileName = appendTimestampToFilename(filename);
                    return [3 /*break*/, 7];
                case 3: return [4 /*yield*/, inquirer.prompt([{
                            type: 'input',
                            name: 'filename',
                            message: 'New name for your file:',
                        }])];
                case 4:
                    ANS_REN = _b.sent();
                    outfileName = ANS_REN.filename;
                    return [3 /*break*/, 7];
                case 5: return [2 /*return*/, filename];
                case 6:
                    console.log('Bye!');
                    process.exit(0);
                    _b.label = 7;
                case 7:
                    OUTPATH = path.join(process.cwd(), dstdir);
                    URI = path.join(OUTPATH, outfileName);
                    if (!fs.existsSync(URI)) return [3 /*break*/, 9];
                    return [4 /*yield*/, userResolveConflictAsync(dstdir, outfileName)];
                case 8: return [2 /*return*/, _b.sent()];
                case 9: return [2 /*return*/, outfileName];
            }
        });
    });
}
exports.userResolveConflictAsync = userResolveConflictAsync;
